checks/permissions.go

```diff
@@ -24,149 +24,235 @@ import (
-	permissionRead := true
-	var r bool
-	var err error
-
+	dl checker.DetailLogger, pdata *permissionCbData) error {
 	// Iterate over the permission, verify keys and values are strings.
 	for k, v := range values {
 		key, ok := k.(string)
 		if !ok {
 			//nolint
-			return false, sce.Create(sce.ErrScorecardInternal, errInvalidGitHubWorkflowFile.Error())
-		}
-
-		if r, err = validatePermission(key, v, path, dl); err != nil {
-			return false, err
+			return sce.Create(sce.ErrScorecardInternal, errInvalidGitHubWorkflowFile.Error())
 		}
 
-		if !r {
-			permissionRead = false
+		if err := validatePermission(key, v, path, dl, pdata); err != nil {
+			return err
 		}
 	}
-	return permissionRead, nil
+	return nil
+}
+
+func recordPermissionWrite(name string, pdata *permissionCbData) {
+	pdata.permissions[name] = true
+}
+
+func recordAllPermissionsWrite(pdata *permissionCbData) {
+	// Special case: `all` does not correspond
+	// to a GitHub permission.
+	pdata.permissions["all"] = true
 }
 
 func validateReadPermissions(config map[interface{}]interface{}, path string,
-	dl checker.DetailLogger) (bool, error) {
+	dl checker.DetailLogger, pdata *permissionCbData) error {
 	var permissions interface{}
 
 	// Check if permissions are set explicitly.
 	permissions, ok := config["permissions"]
 	if !ok {
 		dl.Warn("no permission defined in %v", path)
-		return false, nil
+		recordAllPermissionsWrite(pdata)
+		return nil
 	}
 
 	// Check the type of our values.
 	switch val := permissions.(type) {
 	// Empty string is nil type.
 	// It defaults to 'none'
 	case nil:
-		dl.Info("permission set to 'none' in %v", path)
+		dl.Info("permissions set to 'none' in %v", path)
 	// String type.
 	case string:
 		if !strings.EqualFold(val, "read-all") && val != "" {
-			dl.Warn("permission set to '%v' in %v", val, path)
-			return false, nil
+			dl.Warn("permissions set to '%v' in %v", val, path)
+			recordAllPermissionsWrite(pdata)
+			return nil
 		}
 		dl.Info("permission set to '%v' in %v", val, path)
 
 	// Map type.
 	case map[interface{}]interface{}:
-		if res, err := validateMapPermissions(val, path, dl); err != nil {
-			return false, err
-		} else if !res {
-			return false, nil
+		if err := validateMapPermissions(val, path, dl, pdata); err != nil {
+			return err
 		}
 
 	// Invalid type.
 	default:
 		//nolint
-		return false, sce.Create(sce.ErrScorecardInternal, errInvalidGitHubWorkflowFile.Error())
+		return sce.Create(sce.ErrScorecardInternal, errInvalidGitHubWorkflowFile.Error())
 	}
 
-	return true, nil
+	return nil
+}
+
+// Calculate the score.
+func calculateScore(result permissionCbData) int {
+	// See list https://github.blog/changelog/2021-04-20-github-actions-control-permissions-for-github_token/.
+	// Note: there are legitimate reasons to use some of the permissions like checks, deployments, etc.
+	// in CI/CD systems https://docs.travis-ci.com/user/github-oauth-scopes/.
+	if _, ok := result.permissions["all"]; ok {
```

Comment:
That's the interesting part I need feedback on

Reply:
@inferno-chromium comments on this function? What do you think of the score calculation?